tests/cases/fourslash/typeAtLocation.ts

```diff
@@ -0,0 +1,9 @@
+/// <reference path="fourslash.ts" />
+
+// @strictNullChecks: true
+
+////function test<T extends string | undefined>(x: T) {
+////    x/**/!;
+////}
+
+verify.typeAtLocation("", "string | undefined"); // TODO: GH#21317: should be "T".
```

Comment:
You should remove the `TODO` (and maybe the PR?), the new type at this position is intended, not a bug; see [my comment.](https://github.com/Microsoft/TypeScript/issues/21317#issuecomment-359657955)

We now consider the type at a location of a generic with a nullable constraint that is nonnull-asserted as having been narrowed to the nonnull part of its constraint (since usually you'd do this to access some member or call on it, like a cast).

Also, I'm pretty sure this is exactly what our `type` baselines test. We knew this changed and intended it.

Reply:
Shouldn't it be `string` and not `string | undefined` then?

Reply:
@weswigham IMO the test is legitimate, because it tests the type **before** the assertion (what the issue is about) and not the result of the assertion.
Though maybe that's already covered by the baselines as you mentioned.
Either way the API issue should be fixed.

Reply:
The type before the assertion is going to be the constraint (`string | undefined`, because it's an apparent position thanks to the assertion), and then after the assertion is applied it should just be `string`. This isn't an API issue, we legitimately interpret the type at that position as just the constraint now, that way the bang after it functions somewhat.